rMD-exp/doc/recordmydesktop.1
.TH "RECORDMYDESKTOP" 1 "13/7/2006" "Linux"


.SH NAME
recordMyDesktop \- record desktop sessions to an Ogg\-Theora\-Vorbis file.


.SH SYNOPSIS

.B        recordmydesktop
[
.B
Options
]^
.B
filename
.br
.br
.SH DESCRIPTION
.PP
        recordMyDesktop produces a file(default out.ogg) that contains a video and audio recording
.br
of a linux desktop session. The default behavior of recording is to mark areas that have changed(through libxdamage)
.br
and update the frame. This behavior can be changed (option
.B
\-\-full\-shots
) to produce a more accurate result
.br
or capture windows that do not generate events on change(windows with accelerated 3d context)
.br
but this will notably increase the workload. In this case, enabling the
.B
\-\-with\-shared
option is recommended
.br
(by default this option is switched on or off at every frame, according to the size of the area to be aquired).
.br
.br
recordMyDesktop doesn't have a commandline interface.
.br
After startup, it can be controled only through the following signals:
.br
.br
.B
SIGUSR1
causes the program to pause if it's currently recording, and vice-versa.
.br
.B
SIGTERM
causes normal termination of the recording.
.br
.B
SIGINT
also causes normal termination.
.br
.B
SIGABRT
terminates the program and removes the specified output file.
.br
.br
.br
A typical scenario of recording can be a command as simple as:
.br
.B
~$ recordmydesktop
.br
which will produce a fullscreen recording named out.ogg
.br
while a command like:
.br
.B
~$ recordmydesktop foo.ogg
.br
will write output to foo.ogg
.br
.br
Since version 0.3, encoding will happen right after the recording finishes.
.br
While this behavior saves a lot of CPU, you can revert to the old one by entering the \-\-on\-the\-fly\-encoding switch.
.br
.br
To specify a region for recording you can type this:
.br
.B
~$ recordmydesktop -x X_pos -y Y_pos -width WIDTH -height HEIGHT -o foo.ogg
.br
where X_pos and Y_pos specify the offset in pixels from the upper left
.br
corner of your screen and WIDTH and HEIGHT the size of the window to be recorded(again in pixels).
.br
If the area extends beyond your current resolution, you will be notified appropriately and nothing will happen.
.br
Notice also, that if any option is entered you have to specify the output file with the \-o switch.
.br
If you try to save under a filename that already exists, the name will be post-fixed with a number (incremented if that name exists already)
.br
.br
.B
To normally end a recording you can press ctl-c.
.br
(which will send a
.B
SIGINT
to the program).
.br
.br
For further manipulation of the end result look at the
.B
OPTIONS
and
.B
NOTES
sections.
.br
.br
.br
.SH EXIT STATUS
0 is success
.br
Non-zero means an error occurred, which is printed in stderr.
.br
.br
The following error codes indicate the nature of the error:
.br
1 Error while parsing the arguments.
.br
2 Initializing the encoder failed(either vorbis or theora).
.br
3 Could not open/configure sound card.
.br
4 Xdamage extension not present.
.br
5 Shared memory extension not present.
.br
6 Xfixes extension not present.
.br
7 XInitThreads failed.
.br
8 No $DISPLAY environment variable and none specified as argument.
.br
9 Cannot connect to Xserver.
.br
10 Color depth is not 24bpp.
.br
11 Improper window specification.
.br
12 Cannot attach shared memory to proccess.
.br
13 Cannot open file for writting.
.br
.br
.SH OPTIONS
.PP
.B
Generic Options:
.br
.TP
.B
    \-h or \-\-help
    Print help summary and exit.
.br
.TP
.B
    \-\-version
    Print program version and exit.
.br
.PP
.br
.B
Image Options:
.br
.TP
.B
    \-windowid id_of_window
    id of window to be recorded.
.br
.TP
.B
    \-display DISPLAY
    Display to connect to.
.br
.TP
.B
    \-x X
    Offset in x direction.
.br
.TP
.B
    \-y Y
    Offset in y direction.
.br
.TP
.B
    \-width N
    Width of recorded window.
.br
.TP
.B
    \-height N
    Height of recorded window.
.br
.TP
.B
.br
.br
.TP
.B
    \-dummy\-cursor color
    Draw a dummy cursor, instead of the normal one.Value of color can be "black" or "white".
.br
.TP
.B
    \-\-no\-cursor
    Disable drawing of the cursor.
.br
.TP
.B
    \-\-with\-shared
    Enable usage of MIT\-shared memory extension at all times.
.br
.TP
.B
    \-\-no\-cond\-shared
    Do not use the MIT\-shared memory extension when aquiring large areas.
.br
.TP
.B
     \-shared\-threshold n
     Threshold over which shared memory is used(default 75).
.br
.TP
.B
    \-\-full\-shots
    Take full screenshot at every frame(Not recomended!).
.br
.TP
.B
    \-\-quick\-subsampling
    Do subsampling of the chroma planes by discarding extra pixels.
.br
.TP
.B
    \-fps N(number>0.0)
    A positive number denoting desired framerate.
.br
.br
.PP
.B
Sound Options:
.br
.TP
.B
    \-channels N(number>0)
    A positive number denoting desired sound channels in recording.
.br
.TP
.B
    \-freq N(number>0)
    A positive number denoting desired sound frequency.
.br
.TP
.B
    \-device SOUND_DEVICE
    Sound device(default hw0:0).
.br
.TP
.B
    \-\-no\-sound
    Do not record sound.
.br
.PP
.br
.B
Encoding Options:
.br
.TP
.B
    \-\-on\-the\-fly\-encoding
    Encode the audio-video data, while recording.
.br
.TP
.B
    \-v_quality n
    A number from 0 to 63 for desired encoded video quality(default 63).
.br
.TP
.B
    \-v_bitrate n
    A number from 45000 to 2000000 for desired encoded video bitrate(default 45000).
.br
.TP
.B
    \-\-drop\-frames
    Allow theora encoder to drop frames(this can significantly lower proccessing times).
.br
.TP
.B
    \-s_quality n
    Desired audio quality(\-1 to 10).
.br
.PP
.br
.B
Misc Options:
.br
.TP
.B
    \-\-no\-wm\-check
    When a 3d compositing window manager is detected the program will function as if the
    \-\-full\-shots and \-\-with\-shared options have been specified. This option disables that
    behavior alltogether.
.br
.TP
.B
    \-\-zero\-compression
    Image data are always cached uncompressed.
.br
.TP
.B
    \-workdir DIR
    Location where a temporary directory will be created to hold project files(default $HOME).
.br
.TP
.B
    \-delay n[H|h|M|m]
    Number of secs(default),minutes or hours before capture starts(number can be float).
.br
.TP
.B
    \-\-overwrite
    If there is already a file with the same name, delete it.
    Default action is to add a number postfix to the new file.
    For example when not specifying a name, if out.ogg exists,
    the new file will be out.ogg.1 and if that exists too, out.ogg.2
    and so on (no ad-infinitum though, more like ad-short-integer...)
.br
.TP
.B
    \-o filename
    Name of recorded video(default out.ogg).



.PP
.br
If no other option is specified, filename can be given without the \-o switch.
.br
.br
.SH USAGE
.TP
.B        recordmydesktop
.br
[\-h| \-\-help| \-\-version| \-delay n[H|h|M|m]| \-windowid id_of_window|
.br
\-display DISPLAY| \-x X| \-y Y|\-width N| \-height N| \-fps N(number>0)| \-\-on\-the\-fly\-encoding|
.br
\-v_quality n| \-s_quality n| \-v_bitrate n| \-\-no\-framedrop| \-dummy\-cursor color|
.br
\-\-no\-cursor| \-freq N(number>0)| \-channels N(number>0)| \-device SOUND_DEVICE|
.br
\-\-no\-sound| \-\-with\-shared| \-\-no\-cond\-shared| \-shared\-threshold n| \-\-full\-shots|
.br
\-\-quick\-subsampling| \-workdir DIR| \-\-zero\-compression| \-\-no\-wm\-check| \-\-overwrite| \-o filename]^filename
.br
.br
.br
.br
.SH ENVIRONMENT
.TP
.B
DISPLAY
.br
Display environment variable, specifying X server to connect to.
.br
.SH NOTES
.br
    Recording a window using the \-windowid option, doesn't track the window itself, but the region that it covers.
.br
Also when using that option the \-x,\-y,\-width and \-height options are relative to the specified window area.
.br
An easy way to find out the id of a window, is by using the
.B
xwininfo
program.
.br
Running a command like :
.br
.B
xwininfo |grep "Window id:"|sed \-e "s/xwininfo\\:\\ Window id:\\ // ;s/\\ \.*//"
.br
will give you only the id of the window(which should look like this: 0x4800005)
.br
More conviniently you can put all that in the command that launches recordMyDesktop like this:
.br
.B
~$recordmydesktop -windowid $(xwininfo |grep "Window id:"|sed \-e "s/xwininfo\\:\\ Window id:\\ // ;s/\\ \.*//" )
.br
.br
    Also, the lower quality you select on a video recording (
.B
-v_quality
option), the highest CPU-power that you will need.
.br
So
.B
if you are doing the encoding on the fly
,it's better to start with default values and manipulate
the end\-result with another program.
.br
An excellent converter is
.B
ffmpeg2theora
, which despite its name is also a theora to theora converter.
.br
Changing the quality of a recordng with it,can be as simple as :
.br
.B
ffmpeg2theora infile.ogg \-v 3 \-a 4 \-o outfile.ogg
.br
It can even perform resizing on the size of the recording, or change the overall duration.
.br
.br
.SH BUGS
Does not record 3d windows, if \-\-full\-shots isn't specified.
.br
Saving 65536 files with the same name, will result in upredictable behavior,
.br
which might manifest as an endless loop, or a segmentation fault.
.br
.SH AUTHORS
John Varouhakis(johnvarouhakis@gmail.com)
.br
.SH SEE ALSO
.BR xwininfo(1)
.br
.BR ffmpeg2theora(1)
.br

rMD-exp/ChangeLog
/*VERSION 0.2.1*/
*Unupdated lines bug fixed.
*Fuzzy cursor at border bug fixed.
*Incorrect display of windows that
extended beyond recording height fixed.
*Size adjustement that resulted in extra
or missing pixels at borders has been corrected.
New adjustment is 8 pixels max,split evenly if possible.
*The option to drop frames has been added.
*The option to do better subsampling has been added.
*Default behavior now is to use shared memory when
size of image gets over 75% of total (or a user set threshold).
*Many other minor bugfixes.


/*VERSION 0.2.2*/
*Changes to the configure script.
*BGR to YUV convertion corrected.

/*VERSION 0.2.3*/
*Problems on big-endian architectures, corrected
(thanks to Marcel Unbehaun for bringing up the problem
and providing the neccessary feedback to resolve it).
*Failure to respect change in channels number, while
opening soundcard, which resulted in crashes, segfaults
and repeated error-messages has been fixed.
*other minor bugfixes

/*VERSION 0.2.4*/
*Behavior on lack of $DISPLAY evironment
variable has been corrected.
*Size adjustment is now quantized at 16 pixels,
in order to avoid complete image distortion,
present at certain resolutions, when using shared memory
(conditional or not).
*Some other minor bugfixes, related mostly to default
values of options.

/*VERSION 0.2.5*/
*Drawing of the correct cursor is now supported through
the Xfixes extension.
*Change of behavior on lack of any Xserver extension and
failure to open sound device. The program now exits and informs
instead of continuing with a different configuration.
*Exit status also indicates the nature of any error that occured.
*Documentation has been updated with exit status explanations.
*Subsampling of the chroma planes is made by averaging, while
pixel discarding is an option(used to be otherwise).


/*VERSION 0.2.6*/
*Behavior on stream desync has been changed.
Now, instead of dropping sound buffers, image
capturing speeds up. That is, waiting for timer to
expire is disabled until sync is achieved.
When not recording sound, this does not apply.
*Theora return values are now properly checked,
to avoid inserting corrupt packages into the stream.
*Major memory leaks have been fixed. These were related
to the rectangle insertion algorithm and discarding of
used sound buffers,
*Segmentation fault on seemingly random occasions has been
found and is now fixed. Reason was lack of thread safety over
libogg calls.


/*VERSION 0.3.0*/
*Implemented cache functionality
(Encoding by default happens right after the recording)
*Corrected damage event handling.
*Saving under the same file-name,no longer deletes the previous one,
instead it saves under a number postfixed file.
*Opening the sound device doesn't block anymore.
*A few new options were added related to caching.
*Updated documentation.
*Removed uneeded bitmap screenshot function.




recordmydesktop/src/recordmydesktop.c
/*********************************************************************************
*                             recordMyDesktop                                    *
**********************************************************************************
*                                                                                *
*             Copyright (C) 2006  John Varouhakis                                *
*                                                                                *
*                                                                                *
*    This program is free software; you can redistribute it and/or modify        *
*    it under the terms of the GNU General Public License as published by        *
*    the Free Software Foundation; either version 2 of the License, or           *
*    (at your option) any later version.                                         *
*                                                                                *
*    This program is distributed in the hope that it will be useful,             *
*    but WITHOUT ANY WARRANTY; without even the implied warranty of              *
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               *
*    GNU General Public License for more details.                                *
*                                                                                *
*    You should have received a copy of the GNU General Public License           *
*    along with this program; if not, write to the Free Software                 *
*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA   *
*                                                                                *
*                                                                                *
*                                                                                *
*    For further information contact me at johnvarouhakis@gmail.com              *
**********************************************************************************/


#include <recordmydesktop.h>


int main(int argc,char **argv){
    ProgData pdata;
    int exit_status=0;
    if(XInitThreads ()==0){
        fprintf(stderr,"Couldn't initialize thread support!\n");
        exit(7);
    }
    DEFAULT_ARGS(&pdata.args);
    if(ParseArgs(argc,argv,&pdata.args)){
        exit(1);
    }
    if(pdata.args.display!=NULL)
        pdata.dpy = XOpenDisplay(pdata.args.display);
    else{
        fprintf(stderr,"No display specified for connection!\n");
        exit(8);
    }
    if (pdata.dpy == NULL) {
        fprintf(stderr, "Cannot connect to X server %s\n",pdata.args.display);
        exit(9);
    }
    else{
        EncData enc_data;
        CacheData cache_data;
        pthread_t   poll_damage_t,
                    image_capture_t,
                    image_encode_t,
                    image_cache_t,
                    sound_capture_t,
                    sound_encode_t,
                    sound_cache_t,
                    flush_to_ogg_t,
                    load_cache_t;
        XShmSegmentInfo shminfo;
        int i;

        QUERY_DISPLAY_SPECS(pdata.dpy,&pdata.specs);
        if(pdata.specs.depth!=24){
            fprintf(stderr,"Only 24bpp color depth mode is currently supported.\n");
            exit(10);
        }
        if(SetBRWindow(pdata.dpy,&pdata.brwin,&pdata.specs,&pdata.args))
            exit(11);


        //check if we are under compiz or beryl,in which case we must enable full-shots
        //and with it use of shared memory.User can override this
        pdata.window_manager=((pdata.args.nowmcheck)?NULL:rmdWMCheck(pdata.dpy,pdata.specs.root));
        if(pdata.window_manager==NULL){
            fprintf(stderr,"Not taking window manager into account.\n");
        }
        //Right now only wm's that I know of performing 3d compositing are beryl and compiz.
        //No, the blue screen in metacity doesn't count :)
        //names can be compiz for compiz and beryl/beryl-co/beryl-core for beryl(so it's strncmp )
        else if(!strcmp(pdata.window_manager,"compiz") || !strncmp(pdata.window_manager,"beryl",5)){
            fprintf(stderr,"\nDetected 3d compositing window manager.\n"
                           "Reverting to full screen capture at every frame.\n"
                           "To disable this check run with --no-wm-check\n"
                           "(though that is not advised, since it will probably produce faulty results).\n\n");
            pdata.args.full_shots=1;
            pdata.args.noshared=0;
            pdata.args.nocondshared=1;
        }

        QueryExtensions(pdata.dpy,&pdata.args,&pdata.damage_event, &pdata.damage_error);




        //init data

        //these are globals, look for them at the header
        frames_total=frames_lost=encoder_busy=capture_busy=0;

        fprintf(stderr,"Initializing...\n");
        MakeMatrices();
        if(pdata.args.have_dummy_cursor){
            pdata.dummy_pointer=MakeDummyPointer(&pdata.specs,16,pdata.args.cursor_color,0,&pdata.npxl);
            pdata.dummy_p_size=16;
        }

        if((pdata.args.noshared))
            pdata.datamain=(char *)malloc(pdata.brwin.nbytes);

        if(pdata.args.noshared)
            pdata.datatemp=(char *)malloc(pdata.brwin.nbytes);
        pdata.rect_root[0]=pdata.rect_root[1]=NULL;
        pthread_mutex_init(&pdata.list_mutex[0],NULL);
        pthread_mutex_init(&pdata.list_mutex[1],NULL);
        pthread_mutex_init(&pdata.sound_buffer_mutex,NULL);
        pthread_mutex_init(&pdata.libogg_mutex,NULL);
        pthread_mutex_init(&pdata.yuv_mutex,NULL);

        pthread_cond_init(&pdata.time_cond,NULL);
        pthread_cond_init(&pdata.pause_cond,NULL);
        pthread_cond_init(&pdata.image_buffer_ready,NULL);
        pthread_cond_init(&pdata.sound_buffer_ready,NULL);
        pthread_cond_init(&pdata.sound_data_read,NULL);
        pdata.list_selector=Paused=Aborted=pdata.avd=0;
        pdata.running=1;
        time_cond=&pdata.time_cond;
        pause_cond=&pdata.pause_cond;
        Running=&pdata.running;

        if((pdata.args.noshared)){
            pdata.image=XCreateImage(pdata.dpy, pdata.specs.visual, pdata.specs.depth, ZPixmap, 0,pdata.datamain,pdata.brwin.rgeom.width,
                        pdata.brwin.rgeom.height, 8, 0);
            XInitImage(pdata.image);
            GetZPixmap(pdata.dpy,pdata.specs.root,pdata.image->data,pdata.brwin.rgeom.x,pdata.brwin.rgeom.y,
                        pdata.brwin.rgeom.width,pdata.brwin.rgeom.height);
        }
        if((!pdata.args.noshared)||(!pdata.args.nocondshared)){
            pdata.shimage=XShmCreateImage (pdata.dpy,pdata.specs.visual,pdata.specs.depth,ZPixmap,pdata.datash,
                         &shminfo, pdata.brwin.rgeom.width,pdata.brwin.rgeom.height);
            shminfo.shmid = shmget (IPC_PRIVATE,
                                    pdata.shimage->bytes_per_line * pdata.shimage->height,
                                    IPC_CREAT|0777);
            shminfo.shmaddr = pdata.shimage->data = shmat (shminfo.shmid, 0, 0);
            shminfo.readOnly = False;
            if(!XShmAttach(pdata.dpy,&shminfo)){
                fprintf(stderr,"Failed to attach shared memory to proccess.\n");
                exit(12);
            }
            XShmGetImage(pdata.dpy,pdata.specs.root,pdata.shimage,pdata.brwin.rgeom.x,pdata.brwin.rgeom.y,AllPlanes);
        }
        if(!pdata.args.nosound){
            pdata.sound_handle=OpenDev(pdata.args.device,&pdata.args.channels,&pdata.args.frequency,&pdata.periodsize,            &pdata.periodtime,&pdata.hard_pause);
            if(pdata.sound_handle==NULL){
                fprintf(stderr,"Error while opening/configuring soundcard %s\nTry running with the --no-sound or specify a correct device.\n",pdata.args.device);
                exit(3);
            }
        }

        if(pdata.args.encOnTheFly)
            InitEncoder(&pdata,&enc_data,0);
        else
            InitCacheData(&pdata,&enc_data,&cache_data);

        for(i=0;i<(pdata.enc_data->yuv.y_width*pdata.enc_data->yuv.y_height);i++)
            pdata.enc_data->yuv.y[i]=0;
        for(i=0;i<(pdata.enc_data->yuv.uv_width*pdata.enc_data->yuv.uv_height);i++){
            pdata.enc_data->yuv.v[i]=pdata.enc_data->yuv.u[i]=127;
        }

        if((pdata.args.nocondshared)&&(!pdata.args.noshared)){
            if(pdata.args.no_quick_subsample){
                UPDATE_YUV_BUFFER_IM_AVG((&pdata.enc_data->yuv),((unsigned char*)pdata.shimage->data),
                (pdata.enc_data->x_offset),(pdata.enc_data->y_offset),
                (pdata.brwin.rgeom.width),(pdata.brwin.rgeom.height));
            }
            else{
                UPDATE_YUV_BUFFER_IM((&pdata.enc_data->yuv),((unsigned char*)pdata.shimage->data),
                (pdata.enc_data->x_offset),(pdata.enc_data->y_offset),
                (pdata.brwin.rgeom.width),(pdata.brwin.rgeom.height));
            }
        }
        else{
            if(pdata.args.no_quick_subsample){
                UPDATE_YUV_BUFFER_IM_AVG((&pdata.enc_data->yuv),((unsigned char*)pdata.image->data),
                (pdata.enc_data->x_offset),(pdata.enc_data->y_offset),
                (pdata.brwin.rgeom.width),(pdata.brwin.rgeom.height));
            }
            else{
                UPDATE_YUV_BUFFER_IM((&pdata.enc_data->yuv),((unsigned char*)pdata.image->data),
                (pdata.enc_data->x_offset),(pdata.enc_data->y_offset),
                (pdata.brwin.rgeom.width),(pdata.brwin.rgeom.height));
            }
        }

        pdata.frametime=(1000000)/pdata.args.fps;

        if(pdata.args.delay>0){
            fprintf(stderr,"Will sleep for %d seconds now.\n",pdata.args.delay);
            sleep(pdata.args.delay);
        }

        /*start threads*/
        if(!pdata.args.full_shots)
            pthread_create(&poll_damage_t,NULL,(void *)PollDamage,(void *)&pdata);
        pthread_create(&image_capture_t,NULL,(void *)GetFrame,(void *)&pdata);
        if(pdata.args.encOnTheFly)
            pthread_create(&image_encode_t,NULL,(void *)EncodeImageBuffer,(void *)&pdata);
        else
            pthread_create(&image_cache_t,NULL,(void *)CacheImageBuffer,(void *)&pdata);

        if(!pdata.args.nosound){
            pthread_create(&sound_capture_t,NULL,(void *)CaptureSound,(void *)&pdata);
            if(pdata.args.encOnTheFly)
                pthread_create(&sound_encode_t,NULL,(void *)EncodeSoundBuffer,(void *)&pdata);
            else
                pthread_create(&sound_cache_t,NULL,(void *)CacheSoundBuffer,(void *)&pdata);
        }
        if(pdata.args.encOnTheFly)
            pthread_create(&flush_to_ogg_t,NULL,(void *)FlushToOgg,(void *)&pdata);


        RegisterCallbacks(&pdata.args);
        fprintf(stderr,"Capturing!\n");

        //wait all threads to finish

        pthread_join(image_capture_t,NULL);
        fprintf(stderr,"Shutting down.");
        //if no damage events have been received the thread will get stuck
        pthread_cond_broadcast(&pdata.image_buffer_ready);
        if(pdata.args.encOnTheFly)
            pthread_join(image_encode_t,NULL);
        else
            pthread_join(image_cache_t,NULL);


        fprintf(stderr,".");
        if(!pdata.args.nosound){
            int *snd_exit;
            pthread_join(sound_capture_t,(void *)(&snd_exit));
            fprintf(stderr,".");

            if(pdata.args.encOnTheFly){
                if(!(*snd_exit))
                    pthread_join(sound_encode_t,NULL);
                else{
                    pthread_cancel(sound_encode_t);
                    exit_status=*snd_exit;
                }
            }
            else{
                if(!(*snd_exit))
                    pthread_join(sound_cache_t,NULL);
                else{
                    pthread_cancel(sound_cache_t);
                    exit_status=*snd_exit;
                }
            }
        }
        else
            fprintf(stderr,"..");

        if(pdata.args.encOnTheFly)
            pthread_join(flush_to_ogg_t,NULL);
        fprintf(stderr,".");

        if(!pdata.args.full_shots)
            pthread_join(poll_damage_t,NULL);


        fprintf(stderr,".");
        if((!pdata.args.noshared)||(!pdata.args.nocondshared)){
            XShmDetach (pdata.dpy, &shminfo);
            shmdt (&shminfo.shmaddr);
            shmctl (shminfo.shmid, IPC_RMID, 0);
        }
        fprintf(stderr,"\n");


        //Now that we are done with recording we cancel the timer
        CancelTimer();

/**               Encoding                          */
        if(!pdata.args.encOnTheFly){
            if(!Aborted){
                fprintf(stderr,"Encoding started!\nThis may take several minutes.\n"
                "Pressing Ctrl-C will cancel the procedure (resuming will not be possible, but\n"
                "any portion of the video, which is already encoded won't be deleted).\n"
                "Please wait...\n");
                pdata.running=1;
                InitEncoder(&pdata,&enc_data,1);
                //load encoding and flushing threads
                if(!pdata.args.nosound){
                    //before we start loading again
                    //we need to free any left-overs
                    while(pdata.sound_buffer!=NULL){
                        free(pdata.sound_buffer->data);
                        pdata.sound_buffer=pdata.sound_buffer->next;
                    }
                }
                pthread_create(&flush_to_ogg_t,NULL,(void *)FlushToOgg,(void *)&pdata);


                //start loading image and audio
                pthread_create(&load_cache_t,NULL,(void *)LoadCache,(void *)&pdata);

                //join and finish
                pthread_join(load_cache_t,NULL);
                fprintf(stderr,"Encoding finished!\nWait a moment please...\n");
                pthread_join(flush_to_ogg_t,NULL);
            }
            fprintf(stderr,"Cleanning up cache...\n");
            if(PurgeCache(pdata.cache_data,!pdata.args.nosound))
                fprintf(stderr,"Some error occured while cleaning up cache!\n");
            fprintf(stderr,"Done!!!\n");
        }
/**@_______________________________________________@*/

        //This can happen earlier, but in some cases it might get stuck.
        //So we must make sure the recording is not wasted.
        XCloseDisplay(pdata.dpy);

        if(Aborted && pdata.args.encOnTheFly){
            if(remove(pdata.args.filename)){
                perror("Error while removing file:\n");
                return 1;
            }
            else{
                fprintf(stderr,"SIGABRT received,file %s removed\n",pdata.args.filename);
                return 0;
            }
        }
        else
            fprintf(stderr,"Goodbye!\n");
    }
    return exit_status;
}















